### PR TITLE
Grenade Education Tools

### PR DESCRIPTION
**Pull request by nkwiatek**
Hey splewis and crew, 

In light of ESEA's announcement of [NADR](https://play.esea.net/news/15588), I thought it might be good to share a fork I worked on back in April that adds features related to grenades to practicemode. 

This set of 11 commits adds the following:

- Visible holograms that appear around the map to show nade type and lineup, which can be teleported to by pressing `+use`, and toggled per-player with `.holo`.

- A "scoring" heuristic that tells you whether you landed or a missed a saved grenade. 
- Storage of detonation locations in map configuration, and a debug method for admins to add detonation data to all saved nades en masse (`sm_fixdetonations`).
- [A "learning" mode](https://streamable.com/c7scmq) (`.learn`) that teleports you throughout the saved nades in the map and challenges you to correctly throw the grenade.

These changes have been running on a private server for a few months with regular playtesting and appear stable. (Let me know if you'd like the ip/pass to test it out.) Note this stack also includes [a change to csgoutils](https://github.com/nkwiatek/csgo-practice-mode/commit/753506ec6d011a1107ee98133d426cd707e9877d).

I tried to follow the conventions set forth in the codebase, although I'm sure not perfectly. :)  

I'm asking you to consider taking this as a pull request to the main repo because I think these features could add value to the community you've built, especially as a DIY alternative to ESEA's recent announcement.

### CODE REVIEW

**Comment by Otoris** [comment]
This is actually insane, nice job Nick.

**Comment by splewis** [comment]
This is really nice - great work, really. I'll try to review shortly, but I expect we'll be able to merge in in.

File: scripting/practicemode.sp [hunk 1]
```diff
@@ -1603,5 +1674,94 @@ public void CSU_OnThrowGrenade(int client, int entity, GrenadeType grenadeType,
...
+  const float grenadeDetonationOrigin[3]
+) {
+  if (client == -1) {
+    // I guess this is possible in some race conditions involving map change or disconnect.
+    return;
+  }
+  if (currentEntity == g_LastGrenadeEntity[client]) {
+    g_LastGrenadeDetonationOrigin[client] = grenadeDetonationOrigin;
+  }
+  if (g_GrenadeDetonationSaveQueue.Length) {
```
Comment (splewis) [inline scripting/practicemode.sp:1696]: I'd prefer being explicit here: `g_GrenadeDetonationSaveQueue.Length > 0` or `!g_GrenadeDetonationSaveQueue.Empty())`

File: scripting/practicemode.sp [hunk 2]
```diff
@@ -1603,5 +1674,94 @@ public void CSU_OnThrowGrenade(int client, int entity, GrenadeType grenadeType,
...
+  Learn_OnGrenadeExplode(client, currentEntity, grenade, grenadeDetonationOrigin);
 }
+
+public void CSU_OnManagedGrenadeExplode(
+  int client,
+  int currentEntity, 
+  GrenadeType grenade,
+  const float grenadeDetonationOrigin[3]
+) {
+  if (!g_ManagedGrenadeDetonationsToFix.Size) {
```
Comment (splewis) [inline scripting/practicemode.sp:1731]: Same on avoiding the implicit int->bool conversion.

File: scripting/practicemode/grenade_utils.sp [hunk 3]
```diff
@@ -792,6 +884,64 @@ public Action CorrectGrenadeIdsHelper(const char[] ownerName, const char[] owner
...
+  GrenadeType grenadeType = GrenadeTypeFromString(grenadeTypeStr);
+  if (
+    grenadeType == GrenadeType_None
+    || (grenadeType == GrenadeType_Smoke && g_ManagedGrenadeDetonationsToFixPhase == GRENADE_DETONATION_FIX_PHASE_NONSMOKES)
+    || (grenadeType != GrenadeType_Smoke && g_ManagedGrenadeDetonationsToFixPhase == GRENADE_DETONATION_FIX_PHASE_SMOKES) 
+  ) {
+    return;
+  }
+
+  // TODO: throw in 3 phases to break glass on some maps?
```
Comment (splewis) [inline scripting/practicemode/grenade_utils.sp:928]: Haha, I've never figured out a good way to deal with glass, the todo is good with me.

File: scripting/practicemode/learn.sp [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  ArrayList arr
+) {
+  GrenadeType type = GrenadeTypeFromString(grenadeType);
+  if (type == GrenadeType_None) {
+    return;
+  }
+  for (int i = 0; i < categories.Length; i++) {
+    char buffer[GRENADE_CATEGORY_LENGTH];
+    categories.GetString(i, buffer, sizeof(buffer));
+    if (!strcmp(buffer, LEARN_BLACKLIST_CATEGORY, .caseSensitive=false)) {
```
Comment (splewis) [inline scripting/practicemode/learn.sp:192]: This would be more readable with StrEqual.

File: scripting/practicemode/learn.sp [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+  CreateTimer(LEARN_ADVANCE_DELAY_SEC, _LearnHit_Delay, client);
+}
+
+public Action _LearnHit_Delay(Handle timer, int client) {
+  LearnAdvance(client);
+}
+
+// Used with success or skip; agnostic to intent.
+public void LearnAdvance(const int client) {
+  if (!g_LearnQueue[client].Length) {
```
Comment (splewis) [inline scripting/practicemode/learn.sp:255]: I have negative opinions on too much implicit conversion w.r.t. readability :)

**Comment by splewis** [comment]
Made a quick pass, really only had style nitpicks. Everything looks good - I'm glad this fit nicely in the existing code without too much shuffling. 

I had long ago wanted to store more per-grenade metadata, like the detonation location. I wanted to enable being able to stand in a location, and query for nades that land there, so this definitely makes adding that easier.

**Comment by nkwiatek** [comment]
Apologies for the major delay here; I'll do a pass on feedback this weekend!

**Comment by nkwiatek** [comment]
☝ looks like a resource url is broken here. fwiw i tested this change locally; it should build with a new link to `updater.inc`.

**Comment by danieliser** [comment]
@nkwiatek, @splewis Curious the status of this PR, looks amazing.

That said could this be hooked to a larger database of saved nade throws & meta data?

IE If I build a quick API and storage server where community users could upload lists of their favorite throws and you could then collate and download a list and load it into the game, either manually or more preferably over some type of console command.

I'd be interested to put that together as I haven't found a reliable way to browse and find the nade I really want, so many videos on YT and guides, would be nice to combine them all, could even refine them all as duplicates etc with the right meta.

**Comment by nkwiatek** [comment]
@danieliser 

> Curious the status of this PR, looks amazing.

Since last submitting, I've noticed some instability with multiple people in the server at once where holograms do not appear for all players. I'd like to track that bug down before @splewis merges, which I can get to this weekend.

> That said could this be hooked to a larger database of saved nade throws & meta data?

I don't see why not. You could use Sourcemod's SQL api (`dbi`) to do this, or you could use the cfg format as an interchange format and e.g. curl for it at runtime, or write the .cfg files to disk and distribute sourcemod as a listen server launcher for the player to run locally on their machine.

**Comment by danieliser** [comment]
@nkwiatek that sounds awesome.

**Comment by splewis** [comment]
> That said could this be hooked to a larger database of saved nade throws & meta data?

Yes, I had hoped for such a system long ago. But it's a non-trivial thing to implement, especially since compatibility with the existing local configs needs to be retained. If someone is interested I'd gladly review their plans / PRs.

**Comment by danieliser** [comment]
Just for clarity, I'm working on something now @splewis. First goal is to record nades into configs and automate screencaptures to some degree, dump all that data into a searchable db server with an exposed API.

Second would be to let users submit their own, make playlists, export them to a cfg.

Lastly would be automating that export/import from the API to CSGO practice mode somehow that isnt' super technical.

Got the skillset and resources, just gotta flesh out the concept fully before I dive in.

Not sure how much it directly ties to this library, or if it would be something else entirely that supliments this mod. Either way is fine ultimately.

**Comment by nkwiatek** [comment]
Just pushed a few lingering bugfixes:

- @fc80bbf is basically just addressing log noise.
- @6ce5b07 addresses a silly gotcha in practicemode that probably never came up before: the grenade.cfg `KeyValues ` parser fatals on (escaped) quote chars in the username/description strings etc. Turns out you just need set a flag on the instance, essentially `KeyValues.WorkCorrectly(true)`
- @29dca31 fixes the bug I was hoping to understand: some hologram entities did not persist through round restarts, e.g. as caused by using `.bot`, or a player joining the other team, or `mp_restartgame` etc. Note that I don't know the root cause — why CSGO decides to delete some entities and not others on round restart, or whether there are other cases besides a round restart where CSGO will delete our entities — and so my solution treats the problem by hooking `round_start`.

**Comment by sLickz0rr** [comment]
looks awesome. i just have 0 idea, how to implement this into our server. is there any guide out there?

**Comment by brookslobe** [comment]
Curious as to the feasibility of merging some of these features in, even if only as a fork with a portion of this PR's content.  In particular, I'm entertaining some similar aspirations as Daniel above and am interested in the 3 commits regarding nade detonation locations.

**Comment by nkwiatek** [comment]
@JohnnyJTH thanks for asking. Unfortunately I think the right call is to shelve this. It's a non-trivial featureset that would surely add more surface area for bugs and requests to the project, something that would justify a major point release. I think it's risky to pull without more buy-in as an initiative, especially now that I haven't ran the code myself in some time.

**Pull request closed without merging**